### ISSUE TITLE
Hardware video acceleration does not work on Focal with xserver 1.20.8

### ISSUE THREAD

**Issue by mamarley** [saiarcot895/chromium-ubuntu-build#80]
Starting with xserver 1.20.8 on Focal, the hardware video acceleration on Chromium has stopped working.  The following messages are printed in the debug log:

```
[1693:1693:0409/080658.509268:ERROR:vaapi_wrapper.cc(483)] vaInitialize failed (ignore if using Wayland desktop environment): unknown libva error
[1693:1693:0409/080658.551234:ERROR:vaapi_wrapper.cc(483)] vaInitialize failed (ignore if using Wayland desktop environment): unknown libva error
[1693:1693:0409/080711.997104:ERROR:vaapi_wrapper.cc(1811)] vaPutSurface failed, VA error: the requested function is not implemented
[1693:1693:0409/080711.997346:ERROR:vaapi_video_decode_accelerator.cc(301)] Failed putting surface into pixmap
[1856:14:0409/080711.997557:ERROR:batching_media_log.cc(38)] MediaEvent: {"error":"VDA Error 4"}
```

I determined that this started to occur after the X server was upgraded to 1.20.8 and bisected it down to https://cgit.freedesktop.org/xorg/xserver/commit/?h=server-1.20-branch&id=ad7364d8d7f936b9b08195e47d2f6ee9329ff687.  This effectively enabled the `EGL_MESA_query_driver` extension in X, which causes the Mesa OpenGL driver reported by X to be `iris` instead of `i965`.  It doesn't actually change the driver in use; Mesa has already been using that driver as default for the 20.0 release on supported hardware.  For some reason, that change is causing Chromium to not use hardware video acceleration.  It doesn't break VAAPI in general though; other applications like VLC continue to work correctly.

**Comment by saiarcot895** [comment]
Nice job on the dissection!

So I suspect that VAAPI is still getting initialized, because if it didn't, then there should have been another error message saying `vaInitialize DRM` failed, but then something after that is failing. Can you try these experiements:

1. Run `LIBVA_MESSAGING_LEVEL=2 chromium-browser --enable-logging=stderr --log-level=0 --vmodule=vaapi_wrapper=4,vaapi_video_decode_accelerator=4` to enable a bit more logging
2. Run `LIBVA_DRIVER_NAME=i965 chromium-browser` to test with the older driver.

**Comment by mamarley** [comment]
OK, I think I'm going crazy now.  After I filed the bug report, I decided it would be worth testing the new `iHD` VAAPI driver instead of the old `i965` one.  (In all cases, I have been using the `iris` Mesa driver, not the `i965` Mesa driver.)  The only reason I wasn't using `iHD` to begin with is because it has issues with deinterlacing, but deinterlacing isn't an issue for anything Chromium would be playing, so I figured it was worth a shot.  It worked just fine that way.

So, I went to run it as you specified in your response and I couldn't reproduce the issue any longer, even with the `i965` VAAPI driver.  It now seems to work fine with either VAAPI driver, despite being easily reproduced earlier.  I'm stumped…  The only thing of which I can think is that perhaps some dependency that got installed with the `iHD` driver changed something?  I did have to install `mesa-vdpau-drivers` to silence a warning that was producing in STDERR when loading, but I tried removing that package again and it didn't affect the behavior.

In any case, I realized that what I was trying to do is a corner-case because libva 2.7.0 (which is in Focal) prefers `iHD` if it is installed and supported by the hardware, so most users probably won't run into my scenario because all hardware supported by `iHD` is also supported by the `iris` Mesa driver so the automatic selection logic shouldn't ever use the `i965` VAAPI driver with the `iris` Mesa driver.

It seems I have wasted your time.  I apologize for that.

**Comment by mamarley** [comment]
I figured it out!  The problem seems to be that Chromium tries to use the `iHD` VAAPI driver whenever the `iris` Mesa driver is in use (as reported by `EGL_MESA_query_driver`), regardless of whether or not it is installed instead of falling back to `i965` VAAPI driver in the case that it isn't.  When I installed `iHD` to test it with Chromium, I also set `LIBVA_DRIVER_NAME=i965` in my `.profile` to force other applications to use the `i965` VAAPI driver because of the problems I had mentioned with deinterlacing.  Setting that apparently overrides the default behavior in Chromium, allowing it to work properly even with the `iris` Mesa driver.
